web/auth/auth.go

```diff
@@ -178,59 +184,49 @@ func sessionData(session *sessions.Session) string {
 }
 
 // OAuth2Login tries to authenticate against an oauth2 provider.
-func OAuth2Login(logger *zap.SugaredLogger, db database.Database) echo.HandlerFunc {
-	return func(c echo.Context) error {
-		w := c.Response()
-		r := c.Request()
-
-		provider, err := gothic.GetProviderName(r)
-		if err != nil {
-			logger.Error(err.Error())
-			return echo.NewHTTPError(http.StatusBadRequest, err.Error())
+func OAuth2Login(logger *zap.SugaredLogger, db database.Database, config oauth2.Config) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		if r.Method != "GET" {
+			logger.Errorf("GitHub login failed: request method %s", r.Method)
+			http.Redirect(w, r, "/", http.StatusUnauthorized)
 		}
+		// TODO(vera): adapt to use with other providers if needed
+		provider := "github"
+
+		// TODO(vera): make sure teacher suffix no longer necessary
 		var teacher int
 		if strings.HasSuffix(provider, TeacherSuffix) {
 			teacher = 1
 		}
 		logger.Debugf("Provider: %v ; Teacher: %v", provider, teacher)
-
 		qv := r.URL.Query()
 		logger.Debugf("qv: %v", qv)
-		redirect := extractRedirectURL(r, Redirect)
-		logger.Debugf("redirect: %v", redirect)
+		// redirect := extractRedirectURL(r, Redirect)
+		logger.Debugf("redirect: %v", config.RedirectURL)
 		// TODO: Add a random string to protect against CSRF.
-		qv.Set(State, strconv.Itoa(teacher)+redirect)
-		logger.Debugf("State: %v", strconv.Itoa(teacher)+redirect)
+		qv.Set(State, strconv.Itoa(teacher)+config.RedirectURL)
+		logger.Debugf("State: %v", strconv.Itoa(teacher)+config.RedirectURL)
 		r.URL.RawQuery = qv.Encode()
 		logger.Debugf("RawQuery: %v", r.URL.RawQuery)
-
-		url, err := gothic.GetAuthURL(w, r)
-		if err != nil {
-			logger.Error(err.Error())
-			return echo.NewHTTPError(http.StatusBadRequest, err.Error())
-		}
-		logger.Debugf("Redirecting to %s to perform authentication; AuthURL: %v", provider, url)
-		return c.Redirect(http.StatusTemporaryRedirect, url)
+		logger.Debugf("Redirecting to %s to perform authentication; AuthURL: %v", provider, config.Endpoint.AuthURL)
+		http.Redirect(w, r, config.Endpoint.AuthURL, http.StatusTemporaryRedirect)
 	}
 }
 
 // OAuth2Callback handles the callback from an oauth2 provider.
-func OAuth2Callback(logger *zap.SugaredLogger, db database.Database, scms *Scms) echo.HandlerFunc {
-	return func(c echo.Context) error {
+func OAuth2Callback(logger *zap.SugaredLogger, db database.Database, config oauth2.Config, app *scm.GithubApp, tokens *TokenManager, secret string) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		if r.Method != "GET" {
+			logger.Errorf("GitHub login failed: request method %s", r.Method)
+			http.Redirect(w, r, "/", http.StatusUnauthorized)
```

Comment:
According to the docs for `http.Redirect()`:
*The provided code should be in the 3xx range and is usually StatusMovedPermanently, StatusFound or StatusSeeOther.*

So it is not clear if `http.Redirect()` is the correct approach for 401 `http.StatusUnauthorized`.

Also, when redirecting, does it make sense to continue executing or should the method return.

Lastly, we do a lot of 401 redirects and error logging. I added a helper `unauthorized()` for this... see latest commit.